### ISSUE TITLE
代码错误

### ISSUE THREAD

**Issue by nulloption** [JoaquinChou/SigSiamPlus#2]
![9db5192df0ea0661e8acbf36d832e5b](https://github.com/user-attachments/assets/0d5c7c82-ccff-40ce-a0c7-bcec1e8223b3)
作者这里写的代码是不是有问题，进行无监督对比学习，应该保存无监督分类效果最好的模型参数，而不是最后一次实验的结果

**Comment by JoaquinChou** [comment]
你好，感谢你的提问。无监督对比学习的目的是为了能够获得更好进行特征表示的backbone；基于这个backbone再进行下游的分类或者分割任务。因此，该部分代码只进行特征表示，保存最后一轮的模型，不涉及到下游分类模型的训练。

**Comment by nulloption** [comment]
或者是保留无监督对比损失最小的一轮结果呢

**Comment by JoaquinChou** [comment]
对比损失最小的模型效果不一定好，主要看tsne图上的特征表示。

**Comment by nulloption** [comment]
那么tsne图表示不是看聚类的结果吗？完全在评估上游结果时，加入一个KNN分类器，做无监督分类，来评估最好的模型，微调到下游有监督分类任务

**Comment by JoaquinChou** [comment]
是这样子的，KNN是有监督的机器学习算法，有监督需要用到标签，无法进行无监督的分类。

**Comment by nulloption** [comment]
### eval with SVM ###
        feats_train = [] # tf网络得到的特征向量
        labels_train = [] # true label
        tf_net.eval()

        for i, (x_t,_,x_f,_,label) in enumerate(train_loader):
            # labels = list(map(lambda x: x[0],label.numpy().tolist()))
            labels = label.tolist()
            x_t, x_f = x_t.cuda().contiguous(),x_f.cuda().contiguous()
            with torch.no_grad():
                h_t,h_f,z_t,z_f = tf_net(x_t,x_f)
                z = torch.cat((z_t, z_f), dim=1)
            feats = z.detach().cpu().numpy()

            for feat in feats:
                feats_train.append(feat)
            labels_train += labels

        

        feats_train = np.array(feats_train)
        labels_train = np.array(labels_train)

        feats_test = []
        labels_test = []

        for i, (x_t,x_t_aug,x_f,x_f_aug,label) in enumerate(train_loader):
            # labels = list(map(lambda x: x[0],label.numpy().tolist()))
            labels = label.tolist()
            x_t, x_f = x_t.cuda().contiguous(),x_f.cuda().contiguous()
            with torch.no_grad():
                h_t,h_f,z_t,z_f = tf_net(x_t,x_f)
                z = torch.cat((z_t, z_f), dim=1)
            feats = z.detach().cpu().numpy()
            for feat in feats:
                feats_test.append(feat)
            labels_test += labels
        feats_test = np.array(feats_test)
        labels_test = np.array(labels_test)
        model_tl = SVC(C = 0.01 , kernel ='linear', decision_function_shape='ovr') # 先用训练集训练svm
        # 构建一个使用线性核函数的支持向量机分类器，用于解决多类别分类问题，同时通过参数C控制模型的复杂度和正则化程度
        model_tl.fit(feats_train, labels_train)
        # 模型 model_tl 将会根据给定的训练集数据 feats_train 和相应的标签 labels_train 进行学习，以找到最优的决策边界或者超平面，从而在未见过的数据上进行准确的分类预测。
        test_accuracy = model_tl.score(feats_test, labels_test) # 用测试集得出最好的tf网络
        logging.info(f"Linear Accuracy : {test_accuracy}")

        if test_accuracy > best_accuracy:
            best_accuracy = test_accuracy
            logging.info(f"Saving best...")

**Comment by nulloption** [comment]
您看看可不可以这样来评估

**Comment by nulloption** [comment]
应该是用一个分类器，用训练集的特征向量有监督训练，再用测试集检验结果。是不是可以体现你上游做对比学习的效果呢

**Comment by nulloption** [comment]
这样是不是比保存最后一次训练的模型参数有用呢？

**Comment by JoaquinChou** [comment]
你的想法不错，可以参考下监督对比学习的论文。https://arxiv.org/pdf/2004.11362
此外，如果你认为本工作对你有所启发，麻烦点个star哈，谢谢
